### PR TITLE
 Google custom search (images and text) #524 

### PR DESCRIPTION
**Pull request by functor-soup**
## Goal

To create  `google` commands that is able to search both text and images. 
Users should also be able to customize the search engine via the config file.
User should  be able to put in cli type options while using the `search` and `image` (image search command)
User should be able to set options for search via the appropriate commands 

## Accompilshed
- `google search <query>` returns at-most 10 results (because of custom engine API limitation) with a `title`,`link` and `snippet` for each result.
-   `google image <query>` now searches for images based on what the query is. The image results are displayed in a different order than the vanilla search one due to rendering weirdness on slack
- config is checked for the existence of two keys in existence in addition  to the config-validation function  that exists in `yeti.core` 
- messages are displayed to the user if no results are found
- warning messages are displayed on the server (where the yetibot is running) console in case google returns messages with http failure status codes.
- cli type options can now be put in `search` and `image` commands 
- options can be set via the config file or (inclusive) via the commands on the chat frontend.
- options can be completely rest via the commandline or (inclusive) also be restored to the existing values set in the config file, again via the commands on the frontend.
- options that have been set can be viewed on the chat frontend

## how to test
1. run yetibot locally and have it access slack correctly
2. remove google search engine credentials, restart yetibot and check to see if it logs out that `Google is not configured`
3. Add-in one of the two keys-value pairs ,restart yetibot and verify you see that the message `Google is not configured` is outputted to the console
4. Put in incorrect credentials (for api-key and custom-search-engine-id) and run yetibot and try to enter something like `!google search digimon` (on slack) and you should given the `API credentials are invalid || google died` message. Also check your console for info messages.
5. Put in the correct credentials but keep the `image search` setting disabled and restart yetibot, this time `!google search digimon` should give you results but `!google image digimon` should give you the message `Google image search returned no results!! ... check your search engine settings` 
6. Once you enable the image search part, a query like `!google image ehhdfskjsfdfsdjfhkjh87dhfjksdhjkds` should return the same message as mentioned above ("Google image returned ..... please...)
7. `!google search fshsfdhfshsfdjh8978978sfdjksfdhkh` should return the message `"Google returned no results!!"`
8. Put in in-valid option values in the config file and restart yetibot
9. Validation errors should appear on the terminal output while the yetibot starts up
10. putting in `google show set options` should show `No options have been set`
11. for a list of options that can be set via the chat frontend enter in `google show options`
12. Try getting info on the commands that you can see as a result of entering the above command
13. Try setting options via the `!google set <option> <value>`
14. Try to make sure sure your search results reflect the options you have set
15. Reset these options by entering the command `!google reset options`
16. Now try entering options via option arguments to the search functions and verify the correctness of the search results. 

## for credentials
1. Have a google/gmail account
2. go here https://cse.google.com/cse/all

### CODE REVIEW

**Comment by devth** [comment]
@functor-soup thanks, I'll try to test it this afternoon!

File: src/yetibot/api/google.clj [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+(defn configured? []
+  (let [config-keys   [:api-key :custom-search-engine-id]
+        config-values (select-keys config config-keys)]
+    (and (conf-valid? config)
+       (= (count config-values) 2))))
+
+(defonce api-url "https://www.googleapis.com/customsearch/v1?parameters")
+
+;; this has yet to be used somewhere
+(defonce keywords (atom {}))
```
Comment (devth) [inline src/yetibot/api/google.clj]: I'm curious what the intention is with `keywords` and `accepted-keywords`. Sounds like you wanted to address it in a future PR. That's totally fine.

File: src/yetibot/api/google.clj [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+         (remove nil?)
+         (s/join "\n"))))
+
+;; map a vector of results to a vector
+;; of string representations of the results
+(defn format-results
+  [result-body & {:keys [order] :or {order :normal}}]
+    (let [result (result-body :items)
+          indexed (map vector (range 10) result)
+          format  #(str (first %)
```
Comment (devth) [inline src/yetibot/api/google.clj]: Can we make the numbered list start at 1 instead of 0?

File: src/yetibot/api/google.clj [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+         (vals)
+         (remove nil?)
+         (s/join "\n"))))
+
+;; map a vector of results to a vector
+;; of string representations of the results
+(defn format-results
+  [result-body & {:keys [order] :or {order :normal}}]
+    (let [result (result-body :items)
+          indexed (map vector (range 10) result)
```
Comment (devth) [inline src/yetibot/api/google.clj]: Can use [map-indexed](https://clojuredocs.org/clojure.core/map-indexed) instead.

File: src/yetibot/api/google.clj [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        (map format indexed)))
+
+(defn search
+  ;; main search function,
+  ;; extra refers to map of extra params to the search api
+  ;; order refers to the the way result is display (look
+  ;; at display-order var)
+  [q & {:keys [extra order]
+        :or {extra {} order :normal}}]
+  (info (str "Google search for: " q))
```
Comment (devth) [inline src/yetibot/api/google.clj]: Can simply use `(info "Google search for:" q)`

File: src/yetibot/api/google.clj (hunk 2 above)
Comment (functor-soup) [inline src/yetibot/api/google.clj]: Sounds logical.

File: src/yetibot/api/google.clj (hunk 3 above)
Comment (functor-soup) [inline src/yetibot/api/google.clj]: Ah good for future reference as well. Will do

File: src/yetibot/api/google.clj (hunk 1 above)
Comment (functor-soup) [inline src/yetibot/api/google.clj]: - Oh yeah I thought of having the user set stuff like `!google lang en` or `!google domain google.fr` have them stored in keywords 
- and once the user does a `!google search gintama` there would be call that would add it (the atom map) to  query-params before sending it out
-  There would  would something like `!google reset options` to reset preferences
- The `accepted-keywords` was just to make the code easier to do rather than have `n`
functions do stuff 

I think this is a bit long  and winded. Not sure what I think of doing stuff as you did with jira (?) where you just add flags  example -> `!google search --lang en -w something mars bars`

**Comment by functor-soup** [comment]
@devth While writing the tests, I changed code here and there to make them relatively more loosely coupled and more testable (imo), do take a look and tell me what you think.

File: src/yetibot/api/google.clj [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  [results & {:keys [order] :or {order :normal}}]
+    (let [indexed (map-indexed #(vector (inc %1) %2) results)
+          format  #(str (first %)
+                        ". "
+                        (format-result (second %) :order order)
+                        "\n\n")]
+        (map format indexed)))
+
+(defn search
+  ;; main search function,
```
Comment (devth) [inline src/yetibot/api/google.clj]: Can put these comments in the docstring for `search`.

File: src/yetibot/api/google.clj (hunk 1 above)
Comment (devth) [inline src/yetibot/api/google.clj]: I think the jira case is different. The options used will vary often, whereas you'll almost never change the search language or domain once it's configured. Should we make it an optional config, with defaults?

Comment (functor-soup) [inline src/yetibot/api/google.clj]: @devth config defaults sounds logical. There are other options like safe search and it's degree that I'm pretty sure people switch between, what search index number to start from, the way to sort the results, image sizes ... so not sure.

**Comment by functor-soup** [comment]
@devth not sure why the circle ci tests are failing, they run successfully on my machine.

File: src/yetibot/api/google.clj (hunk 1 above)
Comment (devth) [inline src/yetibot/api/google.clj]: Makes sense. I think for things like safe search, page, sort, image size, etc, jira-style opts would work well. A bunch of overloads for all those options would be pretty confusing and hard to maintain.

Can either:

1. start with simple impl where user's only get to configure `lang` (and anything else) in config.edn then maybe add the rest later
2. do 1 and also provide jira-style opts for the rest

Thoughts?

**Comment by devth** [comment]
No worries about Circle CI – it's caused by a non-deterministic test of mine.
I just started recently playing with Circle, and anyway, Travis passed :)

File: src/yetibot/api/google.clj (hunk 1 above)
Comment (functor-soup) [inline src/yetibot/api/google.clj]: 2 sounds nicer. Way nicer.

**Comment by functor-soup** [comment]
@devh this(the custom options stuff) is going to take a while so sit tight. Might as well finish this feature in one go than have it ooze out into another issue.

**Comment by devth** [comment]
@functor-soup no rush! Excited to see what you come up with.

**Comment by functor-soup** [comment]
@devth Question. If you look at the bottom of the api file of this feature `api/google.clj`, you'll notice this huge map structure, I was wondering if you had a problem if this got shifted to a(an?) `edn` file. If yes, ideally where would this file go in terms of location?

**Comment by devth** [comment]
As large as it is, I'd rather keep the map inline:

- less pieces to keep track
- no edn loading required

**Comment by functor-soup** [comment]
@devth my apologies for the ignorance, but when you say inline do you mean

``` { :wilfred "show"  :supernatural "another show" }```

rather than 

```
{:wilfred "show"
 :supernatural ""another show}

Also if you mean the former, wouldn't that violate the whole nothing more than 80 characters rule?

**Comment by devth** [comment]
@functor-soup I just meant inline as in the .clj file rather than an external .edn. I didn't mean on a single line. The way you have it specified currently is ideal :)

**Comment by devth** [comment]
@functor-soup anything left or are you ready for merge? If you're ready I'll take one more pass on reviewing.

**Comment by functor-soup** [comment]
@devth My apologies, was/am still making changes, will be working on it today, hopefully will cut it out nice and dandy by the end of today.

**Comment by devth** [comment]
@functor-soup no worries! Take your time.

**Comment by functor-soup** [comment]
@devth sorry couldnt work as much last night, lots of overthinking going on here. I'll message you once I get everything in  stating  the stuff is ready for review. I am aware this is dragging on, hopefully will be done soon.

Question:

`Assumption`: the ideal relationship between `command.clj` and `api.clj` is that of a client-server relationship.

The args passed to the commands have to be defined within the `command.clj` file, as this is something the client should decide not the server. But that puts in the problem of coupling as now the server (`command.clj`) also retains some part of the knowledge relevant to the options passed because it contains that huge list. Is it alright to make the `command.clj` file contain that huge  list? but at the same time now it feels like the domain knowledge is being split between the two files. Not sure if I am overthinking this. Any thoughts?

**Comment by devth** [comment]
Agree that options list should reside in the `command`. Parsing options is a command-specific task. `api` doesn't know anything about parsing user input.

**Comment by functor-soup** [comment]
@devth  My apologies for pushing up broken code, will fix this shortly. Wanted to push something up today.

**Comment by functor-soup** [comment]
@devth I'm done with the commits, tests and pr documentation. Please do review this whenever you get a chance

**Comment by devth** [comment]
@functor-soup sorry for the delay. Planning to review this tomorrow afternoon!

**Comment by functor-soup** [comment]
@devth not sure if the ugliness of my code has incapacitated you

**Comment by devth** [comment]
@functor-soup lol not at all :) We had a big [product launch](http://venturebeat.com/2016/10/18/ebay-launches-shopbot-facebook-messenger/) this week and I got too busy! I'm traveling today but I'm aiming to review this tomorrow or Friday. Sorry for the delay!

**Comment by functor-soup** [comment]
@devth sounds insane man. You get to work on a chatbot at work? Insane. 👍  I look forward to the review.

File: src/yetibot/api/google.clj [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+(defonce api-url "https://www.googleapis.com/customsearch/v1?parameters")
+
+(defonce messages
+  {:error-message "Values has to be one of the following: "})
+
+(defonce display-order {:normal [:title :link :snippet]
+                        :image  [:title :snippet :link]})
+
+(def config (config-for-ns))
```
Comment (devth) [inline src/yetibot/api/google.clj]: I'm working on killing all `config-for-ns` usages in another branch as it doesn't reliably determine the correct namespace in all circumstances. Can you pls replace with:

```clojure
(yetibot.core.config/get-config :yetibot :api :google) ;; can refer get-config above
```
?

File: src/yetibot/commands/google.clj [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+      (api/format-results (get-in results [:results :items]) :order :image))))
+
+(defn option-info
+  "google option-info <query> #display info for option"
+  [{[_ query] :match}]
+  (if-let [info (get-info (keyword query))]
+    info
+    (:option-not-found messages)))
+
+(defn show-custom-keywords-list
```
Comment (devth) [inline src/yetibot/commands/google.clj]: fn name should match the command, e.g. `show-options` or `show-options-cmd`. Also, we could just shorten this to `google options` which means the fn would be named `options` or `options-cmd`, and it could return a vector of options so that it would be easy to use with `xargs`. That would allow a user to quickly see all option docs using:


```
!google options | xargs google option-info
```

See [room](https://github.com/devth/yetibot.core/blob/master/src/yetibot/core/commands/room.clj) and [category](https://github.com/devth/yetibot.core/blob/master/src/yetibot/core/commands/category.clj) commands for similar examples / inspiration.

File: src/yetibot/api/google.clj (hunk 6 above)
Comment (functor-soup) [inline src/yetibot/api/google.clj]: - [X] Done

File: src/yetibot/commands/google.clj (hunk 7 above)
Comment (functor-soup) [inline src/yetibot/commands/google.clj]: - [X] done, also realized the output of `!google options | xargs google option-info` was horrendous, so adedd a nicer formatting for it. Do tell me what you think.

**Comment by functor-soup** [comment]
@devth added changes. Hope it's all good.

**Comment by devth** [comment]
@functor-soup thanks!!!! 💯
